client/pd_service_discovery.go

```diff
@@ -1020,32 +1021,41 @@ func (c *pdServiceDiscovery) updateURLs(members []*pdpb.Member) {
 	log.Info("[pd] update member urls", zap.Strings("old-urls", oldURLs), zap.Strings("new-urls", urls))
 }
 
-func (c *pdServiceDiscovery) switchLeader(addrs []string) (bool, error) {
+// switchLeader switches the leader of the PD cluster.
+// Note: For current implementation, when initializing the client, the connection to leader should be established.
+// Otherwise, the initialization will fail.
+func (c *pdServiceDiscovery) switchLeader(addrs []string) (change bool, err error) {
 	// FIXME: How to safely compare leader urls? For now, only allows one client url.
 	addr := addrs[0]
 	oldLeader := c.getLeaderServiceClient()
 	if addr == oldLeader.GetAddress() && oldLeader.GetClientConn() != nil {
 		return false, nil
 	}
 
-	newConn, err := c.GetOrCreateGRPCConn(addr)
+	var newConn *grpc.ClientConn
+	newConn, err = c.GetOrCreateGRPCConn(addr)
 	// If gRPC connect is created successfully or leader is new, still saves.
-	if addr != oldLeader.GetAddress() || newConn != nil {
-		// Set PD leader and Global TSO Allocator (which is also the PD leader)
-		leaderClient := newPDServiceClient(addr, addr, c.tlsCfg, newConn, true)
-		c.leader.Store(leaderClient)
+	if addr != oldLeader.GetAddress() {
+		change = true
```

Comment:
It might fail to connect to the leader?

Reply:
Yes, but the leader does change and it will affect the follower proxy